### ISSUE TITLE
Docker Integration 

### ISSUE THREAD

**Issue by JeremyMarshall** [angular-fullstack/generator-angular-fullstack#1959]
- [ ] I understand that GitHub issues are not for tech support, but for questions specific to this generator, bug reports, and feature requests.

Item | Version
----- | -----
generator-angular-fullstack | 3.7.5
Node | 4.2.2
npm | x.x.x
Operating System | OS X 10 / Windows 10 / Ubuntu 15.10 / etc
etc | etc

Item | Answer
----- | -----
Docker

I merged my changes from my previous Docker PR (closed) into https://github.com/angular-fullstack/generator-angular-fullstack/tree/docker-integration

This does the 
`yo angular-fullstack:docker`
as before.
There is also a git submodule in ci/angular-fullstack-dockerfile pointing back to the docker repo https://github.com/angular-fullstack/angular-fullstack-dockerfile and a script inside it to branch this repo for building in hub.docker. Hopefully this can be done automatically with some CI after a release.
There is work to add the hub.docker automated build as I don't have access to the git repo to create it at https://hub.docker.com/u/generatorangularfullstack/ - This currently points to my personal fork and I don't think it can be changed without deleting and recreating in hub.docker

**Comment by Awk34** [comment]
Please conform to the [Contributing Guidelines](https://github.com/angular-fullstack/generator-angular-fullstack/blob/master/contributing.md).

**Comment by Awk34** [comment]
Do you mind if I tweak your commits?

**Comment by JeremyMarshall** [comment]
not at all. I can take them back and resubmit if you haven't done anything

**Comment by AmirGilboa** [comment]
Can you please explain what is it doing, and why should we use it?
Thanks.

**Comment by Awk34** [comment]
@AmirGilboa https://www.docker.com/

**Comment by JeremyMarshall** [comment]
I moved the submodule and fixed the submodule to use the new path and I forgot to add the version number to the docker FROM statement. It uses the main version number and tracks with it when run

**Comment by JeremyMarshall** [comment]
@AmirGilboa This enhancement is the first step towards deploying a whole application stack (including load-balancers, multiple app servers and databases) to AWS or any other cloud provider. These app servers can be scaled up and down as demand changes
This change in particular uses a standard node.js provided build and tracks the npm changes in the generator to cut down on build times. If apps using the Dockerfile use the https://hub.docker.com/ automated build infrastructure then they will also be rebuilt on upstream changes

**Comment by AmirGilboa** [comment]
Sounds great, thanks for doing this. I just did some reading on Docker, have never used it before. It's a great feature for the generator.

**Comment by corporationsruleyourmom** [comment]
Very interested in seeing this happen as both Openshift and Heroku seem to not be deploying properly!

**Comment by aabm00** [comment]
Yes. it looks that now we can deploy the application to Openshift. Some way to solve this problem?

Thanks

**Comment by JeremyMarshall** [comment]
I added some wiki information.
I have some fixes for circleci waiting to push as I was hoping to update the docker repo from it. Unfortunately it is failing some tests and deployment steps can't go until all the tests pass

**Comment by JeremyMarshall** [comment]
@Awk34 I pushed a fix to the branch to remove grunt and merge master in.

There are a couple of outstanding issues around the docker image generation
* change the github repo backing [docker hub](https://hub.docker.com/r/generatorangularfullstack/angular-fullstack-dist/)
* fix circleci so it can be used to push branches to the github repo backing docker hub - failed builds don't run deployment steps
